irohad/consensus/yac/storage/yac_vote_storage.hpp

```diff
@@ -41,36 +42,79 @@ namespace iroha {
 
-        StorageResult applyCommit(CommitMessage commit,
-                                  uint64_t peers_in_round);
+        /**
+         * Insert commit in storage
+         * @param commit - message with votes
+         * @param peers_in_round - number of peers in current consensus round
+         * @return structure with result of inserting.
+         * Nullopt if commit not valid.
+         */
+        nonstd::optional<Answer> store(CommitMessage commit,
+                                       uint64_t peers_in_round);
 
-        StorageResult applyReject(RejectMessage reject,
-                                  uint64_t peers_in_round);
+        /**
+         * Insert reject message in storage
+         * @param reject - message with votes
+         * @param peers_in_round - number of peers in current consensus round
+         * @return structure with result of inserting.
+         * Nullopt if reject not valid.
+         */
+        nonstd::optional<Answer> store(RejectMessage reject,
+                                       uint64_t peers_in_round);
 
-        nonstd::optional<StorageResult> findProposal(YacHash hash);
+        std::shared_ptr<YacProposalStorage>
+        getProposalStorage(ProposalHash hash);
+        /**
+         * Method provide state of processing for concrete hash
+         * @param hash - target tag
+         * @return value attached to parameter's hash. Default is false.
+         */
+        bool getProcessingState(const ProposalHash &hash);
+
+        /**
+         * Mark hash as processed.
+         * @param hash - target tag
+         */
+        void markAsProcessedState(const ProposalHash &hash);
 
        private:
         // --------| private api |--------
 
         /**
+         * Insert votes in storage
+         * @param votes - collection for insertion
+         * @param peers_in_round - number of peers in current round
+         * @return answer after insertion collection
+         */
+        nonstd::optional<Answer> insert_votes(std::vector<VoteMessage> &votes,
+                                              uint64_t peers_in_round);
+        /**
          * Find existed proposal storage or create new if required
          * @param msg - vote for finding
          * @param peers_in_round - number of peer required
          * for verify supermajority;
          * This parameter used on creation of proposal storage
-         * @return - index of required proposal storage
+         * @return - pointer for required proposal storage
+         */
+        std::shared_ptr<YacProposalStorage>
+        findProposalStorage(const VoteMessage &msg, uint64_t peers_in_round);
+
+        // --------| fields |--------
+
+        /**
+         * Active proposal storages
          */
-        uint64_t findProposalStorage(const VoteMessage &msg,
-                                     uint64_t peers_in_round);
+        std::vector<std::shared_ptr<YacProposalStorage>> proposal_storages_;
```

Comment:
`std::shared_ptr` was used since `YacVoteStorage::getProposalStorage` is public interface. This method is used only as condition for `Yac::votingStep`. 
`YacVoteStorage::getProposalStorage` return value can be replaced with `bool`, and this vector can store `YacProposalStorage` directly.